File: backend/app/models/jobs.py
"""
Pydantic models for Jobs, Labor, Clock-Out Questions, and Daily Reports.

Covers all request/response shapes for Phase 4 endpoints.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# ── Bill Rate Types ───────────────────────────────────────────────────

class BillRateTypeCreate(BaseModel):
    """Create a new bill rate type."""
    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = None


class BillRateTypeUpdate(BaseModel):
    """Update a bill rate type."""
    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = None
    is_active: bool | None = None


class BillRateTypeResponse(BaseModel):
    """Bill rate type lookup entry."""
    id: int
    name: str
    description: str | None = None
    sort_order: int = 0
    is_active: bool = True
    created_at: str | None = None


# ── Jobs ──────────────────────────────────────────────────────────────

class JobCreate(BaseModel):
    """Create a new job."""
    job_number: str
    job_name: str
    customer_name: str
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    state: str | None = None
    zip: str | None = None
    gps_lat: float | None = None
    gps_lng: float | None = None
    status: str = "active"
    priority: str = "normal"
    job_type: str = "service"
    bill_rate_type_id: int | None = None
    lead_user_id: int | None = None
    start_date: str | None = None
    due_date: str | None = None
    notes: str | None = None
    # On Call / Warranty sub-type fields
    on_call_type: str | None = None  # 'on_call' or 'warranty'
    warranty_start_date: str | None = None
    warranty_end_date: str | None = None


class JobUpdate(BaseModel):
    """Update an existing job. All fields optional."""
    job_name: str | None = None
    customer_name: str | None = None
    status: str | None = None
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    state: str | None = None
    zip: str | None = None
    gps_lat: float | None = None
    gps_lng: float | None = None
    priority: str | None = None
    job_type: str | None = None
    bill_rate_type_id: int | None = None
    lead_user_id: int | None = None
    start_date: str | None = None
    due_date: str | None = None
    notes: str | None = None
    # On Call / Warranty sub-type fields
    on_call_type: str | None = None
    warranty_start_date: str | None = None
    warranty_end_date: str | None = None


class JobStatusUpdate(BaseModel):
    """Change job status (separate from general update for audit trail)."""
    # pending, active, on_hold, completed, cancelled,
    # continuous_maintenance, on_call
    status: str


class JobResponse(BaseModel):
    """Full job detail response."""
    id: int
    job_number: str
    job_name: str
    customer_name: str
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    state: str | None = None
    zip: str | None = None
    gps_lat: float | None = None
    gps_lng: float | None = None
    status: str
    priority: str
    job_type: str
    bill_rate_type_id: int | None = None
    bill_rate_type_name: str | None = None
    lead_user_id: int | None = None
    lead_user_name: str | None = None
    start_date: str | None = None
    due_date: str | None = None
    completed_date: str | None = None
    notes: str | None = None
    # On Call / Warranty sub-type fields
    on_call_type: str | None = None
    warranty_start_date: str | None = None
    warranty_end_date: str | None = None
    warranty_days_remaining: int | None = None  # computed by service layer
    created_at: str | None = None
    updated_at: str | None = None
    # Aggregated stats (populated by service layer)
    total_labor_hours: float | None = None
    total_parts_cost: float | None = None
    active_workers: int | None = None
    # Notebook task aggregation
    open_task_count: int = 0
    task_summary: dict | None = None

    # Formatted address for display
    @property
    def full_address(self) -> str | None:
        parts = [
            self.address_line1,
            self.address_line2,
            ", ".join(filter(None, [self.city, self.state])),
            self.zip,
        ]
        addr = " ".join(filter(None, parts))
        return addr if addr.strip() else None


class JobListItem(BaseModel):
    """Lightweight job for list views."""
    id: int
    job_number: str
    job_name: str
    customer_name: str
    address_line1: str | None = None
    city: str | None = None
    state: str | None = None
    zip: str | None = None
    gps_lat: float | None = None
    gps_lng: float | None = None
    status: str
    priority: str
    job_type: str
    bill_rate_type_name: str | None = None
    lead_user_name: str | None = None
    # On Call / Warranty sub-type (for displaying sub-type in lists)
    on_call_type: str | None = None
    warranty_end_date: str | None = None
    active_workers: int = 0
    total_labor_hours: float = 0
    total_parts_cost: float = 0
    open_task_count: int = 0
    created_at: str | None = None


# ── Labor Entries ─────────────────────────────────────────────────────

class ClockInRequest(BaseModel):
    """Clock in to a job."""
    gps_lat: float | None = None
    gps_lng: float | None = None
    # photo_path handled via file upload separately


class ClockOutRequest(BaseModel):
    """Clock out — includes GPS, optional photo, and question responses."""
    labor_entry_id: int
    gps_lat: float | None = None
    gps_lng: float | None = None
    drive_time_minutes: int = 0
    notes: str | None = None
    # Responses to global clock-out questions
    responses: list[ClockOutResponseInput] = Field(default_factory=list)
    # Answers to one-time questions
    one_time_answers: list[OneTimeAnswerInput] = Field(default_factory=list)


class ClockOutResponseInput(BaseModel):
    """Single answer to a global clock-out question."""
    question_id: int
    answer_text: str | None = None
    answer_bool: bool | None = None
    # photo_path handled via file upload


class OneTimeAnswerInput(BaseModel):
    """Answer to a one-time per-job question."""
    question_id: int
    answer_text: str | None = None
    # photo_path handled via file upload


class LaborEntryResponse(BaseModel):
    """Labor entry with user and job context."""
    id: int
    user_id: int
    user_name: str | None = None
    job_id: int
    job_name: str | None = None
    job_number: str | None = None
    clock_in: str
    clock_out: str | None = None
    regular_hours: float | None = None
    overtime_hours: float | None = None
    drive_time_minutes: int = 0
    clock_in_gps_lat: float | None = None
    clock_in_gps_lng: float | None = None
    clock_out_gps_lat: float | None = None
    clock_out_gps_lng: float | None = None
    clock_in_photo_path: str | None = None
    clock_out_photo_path: str | None = None
    status: str
    notes: str | None = None
    created_at: str | None = None


class ActiveClockResponse(BaseModel):
    """Current user's active clock-in, if any."""
    is_clocked_in: bool
    entry: LaborEntryResponse | None = None


# ── Clock-Out Questions ───────────────────────────────────────────────

class ClockOutQuestionCreate(BaseModel):
    """Create or update a global clock-out question."""
    question_text: str
    answer_type: str = "text"  # text, yes_no, photo
    is_required: bool = True
    sort_order: int = 0


class ClockOutQuestionResponse(BaseModel):
    """Global clock-out question definition."""
    id: int
    question_text: str
    answer_type: str
    is_required: bool
    sort_order: int
    is_active: bool
    created_at: str | None = None


class QuestionReorderRequest(BaseModel):
    """Reorder questions by providing ordered list of IDs."""
    ordered_ids: list[int]


# ── One-Time Questions ────────────────────────────────────────────────

class OneTimeQuestionCreate(BaseModel):
    """Boss creates a one-time question for a specific job/worker."""
    question_text: str
    answer_type: str = "text"
    target_user_id: int | None = None  # NULL = ask everyone


class OneTimeQuestionResponse(BaseModel):
    """One-time question with current status."""
    id: int
    job_id: int
    target_user_id: int | None = None
    target_user_name: str | None = None
    question_text: str
    answer_type: str
    status: str
    created_by: int
    created_by_name: str | None = None
    answered_by: int | None = None
    answer_text: str | None = None
    answer_photo_path: str | None = None
    created_at: str | None = None
    answered_at: str | None = None


# ── Clock-Out Bundle ──────────────────────────────────────────────────

class ClockOutBundle(BaseModel):
    """Everything needed for the clock-out flow UI.

    Combines global questions + pending one-time questions into a single
    payload so the frontend can render the full questionnaire in one fetch.
    """
    global_questions: list[ClockOutQuestionResponse] = Field(default_factory=list)
    one_time_questions: list[OneTimeQuestionResponse] = Field(default_factory=list)


# ── Job Parts ─────────────────────────────────────────────────────────

class JobPartConsumeRequest(BaseModel):
    """Record parts consumed on a job."""
    part_id: int
    qty_consumed: int
    notes: str | None = None


class JobPartResponse(BaseModel):
    """Part consumed on a job."""
    id: int
    job_id: int
    part_id: int
    part_name: str | None = None
    part_code: str | None = None
    qty_consumed: int
    qty_returned: int
    unit_cost_at_consume: float | None = None
    unit_sell_at_consume: float | None = None
    consumed_by: int | None = None
    consumed_by_name: str | None = None
    consumed_at: str | None = None
    notes: str | None = None


# ── Daily Reports ─────────────────────────────────────────────────────

class DailyReportResponse(BaseModel):
    """Daily report metadata (without the full JSON blob)."""
    id: int
    job_id: int
    job_name: str | None = None
    job_number: str | None = None
    report_date: str
    status: str
    generated_at: str | None = None
    reviewed_by: int | None = None
    reviewed_at: str | None = None
    # Summary fields extracted from report_json
    worker_count: int = 0
    total_labor_hours: float = 0
    total_parts_cost: float = 0


class DailyReportFull(BaseModel):
    """Full daily report including the rendered JSON data."""
    id: int
    job_id: int
    job_name: str | None = None
    job_number: str | None = None
    report_date: str
    status: str
    generated_at: str | None = None
    report_data: dict[str, Any] = Field(default_factory=dict)

File: src/lib/types/jobs.ts
/**
 * Jobs & Labor types — jobs, bill rates, labor entries, clock-out questions, daily reports.
 */

// ═══════════════════════════════════════════════════════════════════
// JOBS MODULE (Phase 4)
// ═══════════════════════════════════════════════════════════════════

// ── Job Types ────────────────────────────────────────────────────

export type JobStatus =
  | 'pending' | 'active' | 'on_hold'
  | 'completed' | 'cancelled'
  | 'continuous_maintenance' | 'on_call';
export type JobPriority = 'low' | 'normal' | 'high' | 'urgent';
export type JobType = 'service' | 'new_construction' | 'remodel' | 'maintenance' | 'emergency';
export type OnCallType = 'on_call' | 'warranty';

/** Display labels for on_call sub-types */
export const ON_CALL_TYPE_LABELS: Record<OnCallType, string> = {
  on_call: 'On Call',
  warranty: 'Warranty',
};

/** Human-readable display labels for job statuses */
export const JOB_STATUS_LABELS: Record<JobStatus, string> = {
  pending: 'Pending',
  active: 'Active',
  on_hold: 'On Hold',
  completed: 'Completed',
  cancelled: 'Cancelled',
  continuous_maintenance: 'Cont. Maint.',
  on_call: 'On Call / Warranty',
};

// ── Bill Rate Types ──────────────────────────────────────────────

export interface BillRateType {
  id: number;
  name: string;
  description?: string | null;
  sort_order: number;
  is_active: boolean;
  created_at?: string | null;
}

export interface BillRateTypeCreate {
  name: string;
  description?: string | null;
}

export interface BillRateTypeUpdate {
  name?: string;
  description?: string | null;
  is_active?: boolean;
}
export type LaborStatus = 'clocked_in' | 'clocked_out' | 'edited' | 'approved';
export type ReportStatus = 'generated' | 'reviewed' | 'locked';

export interface JobCreate {
  job_number: string;
  job_name: string;
  customer_name: string;
  address_line1?: string;
  address_line2?: string;
  city?: string;
  state?: string;
  zip?: string;
  gps_lat?: number;
  gps_lng?: number;
  status?: JobStatus;
  priority?: JobPriority;
  job_type?: JobType;
  bill_rate_type_id?: number;
  lead_user_id?: number;
  start_date?: string;
  due_date?: string;
  notes?: string;
  on_call_type?: OnCallType;
  warranty_start_date?: string;
  warranty_end_date?: string;
}

export interface JobUpdate {
  job_name?: string;
  customer_name?: string;
  status?: JobStatus;
  address_line1?: string;
  address_line2?: string;
  city?: string;
  state?: string;
  zip?: string;
  gps_lat?: number;
  gps_lng?: number;
  priority?: JobPriority;
  job_type?: JobType;
  bill_rate_type_id?: number;
  lead_user_id?: number;
  start_date?: string;
  due_date?: string;
  notes?: string;
  on_call_type?: OnCallType | null;
  warranty_start_date?: string | null;
  warranty_end_date?: string | null;
}

export interface JobTeamMember {
  id: number;
  job_id: number;
  user_id: number;
  display_name: string;
  email?: string | null;
  role: 'lead' | 'member';
  assigned_at: string;
  notes?: string | null;
}

export interface JobResponse {
  id: number;
  job_number: string;
  job_name: string;
  customer_name: string;
  address_line1?: string | null;
  address_line2?: string | null;
  city?: string | null;
  state?: string | null;
  zip?: string | null;
  gps_lat?: number | null;
  gps_lng?: number | null;
  status: JobStatus;
  priority: JobPriority;
  job_type: JobType;
  bill_rate_type_id?: number | null;
  bill_rate_type_name?: string | null;
  lead_user_id?: number | null;
  lead_user_name?: string | null;
  start_date?: string | null;
  due_date?: string | null;
  completed_date?: string | null;
  notes?: string | null;
  on_call_type?: OnCallType | null;
  warranty_start_date?: string | null;
  warranty_end_date?: string | null;
  warranty_days_remaining?: number | null;
  created_at?: string | null;
  updated_at?: string | null;
  // Aggregated stats
  total_labor_hours?: number | null;
  total_parts_cost?: number | null;
  active_workers?: number | null;
  // Notebook task aggregation
  open_task_count: number;
  task_summary?: Record<string, number> | null;
}

export interface JobListItem {
  id: number;
  job_number: string;
  job_name: string;
  customer_name: string;
  address_line1?: string | null;
  city?: string | null;
  state?: string | null;
  zip?: string | null;
  gps_lat?: number | null;
  gps_lng?: number | null;
  status: JobStatus;
  priority: JobPriority;
  job_type: JobType;
  bill_rate_type_name?: string | null;
  lead_user_name?: string | null;
  on_call_type?: OnCallType | null;
  warranty_end_date?: string | null;
  active_workers: number;
  total_labor_hours: number;
  total_parts_cost: number;
  open_task_count: number;
  created_at?: string | null;
}

// ── Labor Entry Types ────────────────────────────────────────────

export interface ClockInRequest {
  gps_lat?: number;
  gps_lng?: number;
}

export interface ClockOutResponseInput {
  question_id: number;
  answer_text?: string | null;
  answer_bool?: boolean | null;
}

export interface OneTimeAnswerInput {
  question_id: number;
  answer_text?: string | null;
}

export interface ClockOutRequest {
  labor_entry_id: number;
  gps_lat?: number;
  gps_lng?: number;
  drive_time_minutes?: number;
  notes?: string;
  responses: ClockOutResponseInput[];
  one_time_answers: OneTimeAnswerInput[];
}

export interface LaborEntryResponse {
  id: number;
  user_id: number;
  user_name?: string | null;
  job_id: number;
  job_name?: string | null;
  job_number?: string | null;
  clock_in: string;
  clock_out?: string | null;
  regular_hours?: number | null;
  overtime_hours?: number | null;
  drive_time_minutes: number;
  clock_in_gps_lat?: number | null;
  clock_in_gps_lng?: number | null;
  clock_out_gps_lat?: number | null;
  clock_out_gps_lng?: number | null;
  clock_in_photo_path?: string | null;
  clock_out_photo_path?: string | null;
  status: LaborStatus;
  notes?: string | null;
  created_at?: string | null;
}

export interface ActiveClockResponse {
  is_clocked_in: boolean;
  entry?: LaborEntryResponse | null;
}

// ── Clock-Out Questions ──────────────────────────────────────────

export type QuestionAnswerType = 'text' | 'yes_no' | 'photo';

export interface ClockOutQuestionResponse {
  id: number;
  question_text: string;
  answer_type: QuestionAnswerType;
  is_required: boolean;
  sort_order: number;
  is_active: boolean;
  created_at?: string | null;
}

export interface ClockOutQuestionCreate {
  question_text: string;
  answer_type?: QuestionAnswerType;
  is_required?: boolean;
  sort_order?: number;
}

// ── One-Time Questions ───────────────────────────────────────────

export type OneTimeQuestionStatus = 'pending' | 'answered' | 'expired' | 'cancelled';

export interface OneTimeQuestionResponse {
  id: number;
  job_id: number;
  target_user_id?: number | null;
  target_user_name?: string | null;
  question_text: string;
  answer_type: QuestionAnswerType;
  status: OneTimeQuestionStatus;
  created_by: number;
  created_by_name?: string | null;
  answered_by?: number | null;
  answer_text?: string | null;
  answer_photo_path?: string | null;
  created_at?: string | null;
  answered_at?: string | null;
}

export interface OneTimeQuestionCreate {
  question_text: string;
  answer_type?: QuestionAnswerType;
  target_user_id?: number | null;
}

// ── Clock-Out Bundle ─────────────────────────────────────────────

export interface ClockOutBundle {
  global_questions: ClockOutQuestionResponse[];
  one_time_questions: OneTimeQuestionResponse[];
}

// ── Job Parts ────────────────────────────────────────────────────

export interface JobPartConsumeRequest {
  part_id: number;
  qty_consumed: number;
  notes?: string;
}

export interface JobPartResponse {
  id: number;
  job_id: number;
  part_id: number;
  part_name?: string | null;
  part_code?: string | null;
  qty_consumed: number;
  qty_returned: number;
  unit_cost_at_consume?: number | null;
  unit_sell_at_consume?: number | null;
  consumed_by?: number | null;
  consumed_by_name?: string | null;
  consumed_at?: string | null;
  notes?: string | null;
}

// ── Daily Reports ────────────────────────────────────────────────

export interface DailyReportResponse {
  id: number;
  job_id: number;
  job_name?: string | null;
  job_number?: string | null;
  report_date: string;
  status: ReportStatus;
  generated_at?: string | null;
  reviewed_by?: number | null;
  reviewed_at?: string | null;
  // Summary fields extracted from report JSON
  worker_count: number;
  total_labor_hours: number;
  total_parts_cost: number;
  total_drive_time_minutes?: number;
}

export interface DailyReportFull {
  id: number;
  job_id: number;
  job_name?: string | null;
  job_number?: string | null;
  report_date: string;
  status: ReportStatus;
  generated_at?: string | null;
  report_data: ReportData;
}

// ── Report Data (the JSON blob structure) ────────────────────────

export interface ReportData {
  job_id: number;
  job_name: string;
  job_number: string;
  report_date: string;
  workers: ReportWorker[];
  parts_consumed: ReportPartConsumed[];
  deliveries?: ReportDelivery[];
  trip_legs?: ReportTripLeg[];
  vehicles_involved?: ReportVehicleInvolved[];
  summary: ReportSummary;
}

export interface ReportWorker {
  user_id: number;
  display_name: string;
  clock_in: string;
  clock_out?: string | null;
  regular_hours: number;
  overtime_hours: number;
  drive_time_minutes: number;
  clock_in_gps?: { lat: number; lng: number } | null;
  clock_out_gps?: { lat: number; lng: number } | null;
  clock_in_photo?: string | null;
  clock_out_photo?: string | null;
  responses: ReportQuestionAnswer[];
  one_time_responses: ReportOneTimeAnswer[];
}

export interface ReportQuestionAnswer {
  question: string;
  type: QuestionAnswerType;
  answer: string | boolean;
  photo?: string | null;
}

export interface ReportOneTimeAnswer {
  question: string;
  answer: string;
  photo?: string | null;
}

export interface ReportPartConsumed {
  part_name: string;
  part_code?: string | null;
  qty: number;
  unit_cost: number;
  total: number;
}

export interface ReportSummary {
  total_labor_hours: number;
  total_parts_cost: number;
  worker_count: number;
  total_delivery_items?: number;
  total_miles_driven?: number;
  total_billable_drive_minutes?: number;
  vehicles_involved_count?: number;
}

export interface ReportDelivery {
  vehicle_name: string;
  vehicle_number: string;
  part_name: string;
  part_code?: string | null;
  qty_delivered: number;
  delivered_by_name?: string | null;
  delivered_at?: string | null;
}

export interface ReportTripLeg {
  leg_type: string;
  from_label?: string | null;
  to_label?: string | null;
  miles?: number | null;
  drive_minutes?: number | null;
  is_billable: boolean;
  vehicle_name: string;
  vehicle_number: string;
  driver_name: string;
}

export interface ReportVehicleInvolved {
  vehicle_name: string;
  vehicle_number: string;
  drivers: string[];
  total_miles: number;
  delivered_items: number;
}
